### PR TITLE
core, core/types: regenerate JSON marshaling, add "hash" to headers

### LINKED ISSUE #13858: eth/filters: JSON headers from newHeads sub should include hash
#### System information

Geth version: `Version: 1.6.0-unstable`
OS & Version: Windows 10
Commit hash : `288700c4d884321b7de7119b596c6e052fa33b27`

#### Expected behaviour

`newHeads` subscription returns blocks with `hash`, `size`, `totalDifficulty`, `transactions`, and `uncles` properties.

#### Actual behaviour

`newHeads` subscription returns blocks that are missing those 4 properties.

#### Steps to reproduce the behaviour

1. Launch geth with `geth --testnet --rpc --rpcport 8545 --ws --wsport 8546 --wsorigins "*" --wsapi "eth,net"`
2. Open a websocket connection.
3. subscribe to new heads with `method: "eth_subscribe"`, params: ["newHeads"]`
4. Wait for new block.
5. Notice that the block received over the subscription is missing the mentioned 4 properties.

### PR DESCRIPTION
**Pull request by fjl**
This PR uses `gencodec:"required"` instead of `optional:"true"`, following an upstream change in gencodec. It also adds the hash to marshaled headers. 

Fixes #13858.

### CODE REVIEW

File: core/types/gen_tx_json.go [hunk 1]
```diff
@@ -12,19 +12,19 @@ import (
...
+		AccountNonce hexutil.Uint64  `json:"nonce"    gencodec:"required"`
+		Price        *hexutil.Big    `json:"gasPrice" gencodec:"required"`
+		GasLimit     *hexutil.Big    `json:"gas"      gencodec:"required"`
+		Recipient    *common.Address `json:"to"       rlp:"nil"`
+		Amount       *hexutil.Big    `json:"value"    gencodec:"required"`
+		Payload      hexutil.Bytes   `json:"input"    gencodec:"required"`
+		V            *hexutil.Big    `json:"v" gencodec:"required"`
+		R            *hexutil.Big    `json:"r" gencodec:"required"`
+		S            *hexutil.Big    `json:"s" gencodec:"required"`
 		Hash         *common.Hash    `json:"hash" optional:"yes" rlp:"-"`
```
Comment (karalabe) [inline core/types/gen_tx_json.go]: Shouldn't this `optional:"yes"` disappear?

File: core/types/transaction.go [hunk 2]
```diff
@@ -55,17 +55,17 @@ type Transaction struct {
...
 	// Signature values
-	V *big.Int `json:"v"`
-	R *big.Int `json:"r"`
-	S *big.Int `json:"s"`
+	V *big.Int `json:"v" gencodec:"required"`
+	R *big.Int `json:"r" gencodec:"required"`
+	S *big.Int `json:"s" gencodec:"required"`
 
 	// This is only used when marshaling to JSON.
 	Hash *common.Hash `json:"hash" optional:"yes" rlp:"-"`
```
Comment (karalabe) [inline core/types/transaction.go]: Shouldn't this optional:"yes" disappear?

File: core/types/gen_tx_json.go [hunk 3]
```diff
@@ -39,61 +39,59 @@ func (t txdata) MarshalJSON() ([]byte, error) {
...
+		AccountNonce *hexutil.Uint64 `json:"nonce"    gencodec:"required"`
+		Price        *hexutil.Big    `json:"gasPrice" gencodec:"required"`
+		GasLimit     *hexutil.Big    `json:"gas"      gencodec:"required"`
+		Recipient    *common.Address `json:"to"       rlp:"nil"`
+		Amount       *hexutil.Big    `json:"value"    gencodec:"required"`
+		Payload      hexutil.Bytes   `json:"input"    gencodec:"required"`
+		V            *hexutil.Big    `json:"v" gencodec:"required"`
+		R            *hexutil.Big    `json:"r" gencodec:"required"`
+		S            *hexutil.Big    `json:"s" gencodec:"required"`
 		Hash         *common.Hash    `json:"hash" optional:"yes" rlp:"-"`
```
Comment (karalabe) [inline core/types/gen_tx_json.go]: Shouldn't this optional:"yes" disappear?

File: core/types/gen_tx_json.go (hunk 1 above)
Comment (fjl) [inline core/types/gen_tx_json.go]: yes, will remove

**Comment by fjl** [comment]
@karalabe PTAL

File: core/types/gen_header_json.go [hunk 4]
```diff
@@ -12,24 +12,25 @@ import (
...
+		ReceiptHash common.Hash    `json:"receiptsRoot"     gencodec:"required"`
+		Bloom       Bloom          `json:"logsBloom"        gencodec:"required"`
+		Difficulty  *hexutil.Big   `json:"difficulty"       gencodec:"required"`
+		Number      *hexutil.Big   `json:"number"           gencodec:"required"`
+		GasLimit    *hexutil.Big   `json:"gasLimit"         gencodec:"required"`
+		GasUsed     *hexutil.Big   `json:"gasUsed"          gencodec:"required"`
+		Time        *hexutil.Big   `json:"timestamp"        gencodec:"required"`
+		Extra       hexutil.Bytes  `json:"extraData"        gencodec:"required"`
+		MixDigest   common.Hash    `json:"mixHash"          gencodec:"required"`
+		Nonce       BlockNonce     `json:"nonce"            gencodec:"required"`
```
Comment (karalabe) [inline core/types/gen_header_json.go]: Is there any reason why all the `gencodec:"required"` tags are preserved within the generated code? As I understand it correctly, these fields are used to signal to the unmarshal that a missing field should cause an error. However that is "hard coded" in the generated unmarshal method, so I don't see why we'd need to preserve the tags in the generated code too.

Comment (fjl) [inline core/types/gen_header_json.go]: All tag strings are copied over into the intermediate marshaling type. gencodec parses the tag using package reflect and doesn't know anything about the format. Removing the tags would mean that gencodec needs to construct a new tag with the `gencodec:"..."` part removed. Feel free to open a PR that does this if the extra tags bother you, but there is zero benefit :)

Comment (karalabe) [inline core/types/gen_header_json.go]: Well, since you put it that way. LGTM :+1:  :D

**Pull request merged**